## NumPy

NumPy는 "Numerical Python"의 약자라고도 한다. 이름이 의미하듯이 파이썬의 수치연산 기능을 강화시켜준다.

### Array

NumPy는 자체적인 array를 제공한다. Python의 list와 비슷하지만, 보다 다양한 연산 메소드를 가지고 있다는 점, 같은 종류의 데이터만 담을 수 있다는 점에서 다르다. NumPy의 array는 다음과 같이 사용한다.

In [1]:
import numpy as np

In [2]:
array1 = np.array([1, 2, 3])   # 1차원 array를 만든다.
print(array1.shape)            # (3,) 이라고 출력. 요소가 3인 1차원 array를 의미함          
print(array1)

(3,)
[1 2 3]


In [3]:
# 특정 요소에 접근하기 위해서는 list와 같이 인덱스를 사용한다.
print(array1[0], array1[2])

array1[1] = 10
print(array1)

1 3
[ 1 10  3]


In [4]:
array2 = np.array([[1,2,3],[4,5,6]])   # 2차원 array를 만든다.
print(array2.shape)                    # (2, 3) 이라고 출력. (row, column)
print(array2)
print(array2[0, 0], array2[0, 1], array2[1, 0])   # Prints "1 2 4"

(2, 3)
[[1 2 3]
 [4 5 6]]
1 2 4


In [5]:
a = np.zeros((2,3))   # 모두 zero로 채워진 2x3 array를 생성
print(a)

b = np.ones((1,2))    # 모두 1로 채워진 1x2 array를 생성
print(b)

c = np.random.random((2,2)) # 랜덤 넘버로 채워진 2x2 array를 생성 (Return random floats in the half-open interval [0.0, 1.0))
print(c)

[[0. 0. 0.]
 [0. 0. 0.]]
[[1. 1.]]
[[0.13953938 0.00533994]
 [0.87476191 0.24838326]]


In [6]:
a = np.array([[1,2,3,4], [5,6,7,8], [9,10,11,12]])
print(a)

b = a[:2, 1:3]
print(b)

[[ 1  2  3  4]
 [ 5  6  7  8]
 [ 9 10 11 12]]
[[2 3]
 [6 7]]


In [7]:
a1 = a[1, :]
a2 = a[:2, :]
print(a1)
print(a2)

[5 6 7 8]
[[1 2 3 4]
 [5 6 7 8]]


### Array Datatype

Numpy는 다양한 데이터타입을 제공한다. 데이터타입은 array를 생성할 때 지정할 수 있다.

* bool_  : Boolean (True or False) stored as a byte
* int_  : Default integer type (Int32 또는 Int64 와 같음)
* int8  : Byte (-128 to 127)
* int16  : Integer (-32768 to 32767)
* int32  : Integer (-2147483648 to 2147483647)
* int64  : Integer (-9223372036854775808 to 9223372036854775807)
* uint8  : Unsigned integer (0 to 255)
* uint16 : Unsigned integer (0 to 65535)
* uint32 : Unsigned integer (0 to 4294967295)
* uint64  : Unsigned integer (0 to 18446744073709551615)
* float_ : float64 와 같음
* float16 : Half precision float
* float32 : Single precision float
* float64 : Double precision float
* complex_ : complex128 와 같음
* complex64 : Complex number, represented by two 32-bit floats
* complex128 : Complex number, represented by two 64-bit floats  
* object_  : python object
* str_  : python str
* unicode_  : python unicode

In [8]:
x = np.array([1, 2])      # default로 datatype을 int64로 지정한다.
print(x.dtype)
print(x)

x = np.array([1.0, 2.0])  # default로 datatype을 float64로 지정한다.
print(x.dtype)
print(x)

x = np.array([1, 2], dtype=np.float64)  # array를 생성할 때 datatype을 float64로 지정한다.
print(x.dtype)
print(x)

int64
[1 2]
float64
[1. 2.]
float64
[1. 2.]


### Array Math

In [9]:
x = np.array([[1,2],[3,4]], dtype=np.float64)
y = np.array([[5,6],[7,8]], dtype=np.float64)
print(x)
print(y)

[[1. 2.]
 [3. 4.]]
[[5. 6.]
 [7. 8.]]


In [10]:
print(x + y)
print(np.add(x, y))

[[ 6.  8.]
 [10. 12.]]
[[ 6.  8.]
 [10. 12.]]


In [11]:
print(x - y)
print(np.subtract(x, y))

[[-4. -4.]
 [-4. -4.]]
[[-4. -4.]
 [-4. -4.]]


In [12]:
print(x * y)
print(np.multiply(x, y))

[[ 5. 12.]
 [21. 32.]]
[[ 5. 12.]
 [21. 32.]]


In [13]:
print(np.sqrt(x))

[[1.         1.41421356]
 [1.73205081 2.        ]]


**(참고) 행렬의 곱**

행렬의 곱이 정의되는 경우
 - 행렬 곱 AB에 대하여, 행렬 A의 열의 수와 행렬 B의 행의 수가 같아야 한다.
 
![](img/image1.png)

In [14]:
x = np.array([[1,2],[3,4],[5,6],[7,8]])

print(x)

print(np.mean(x))
print(np.std(x))
print(np.max(x))
print(np.min(x))

[[1 2]
 [3 4]
 [5 6]
 [7 8]]
4.5
2.29128784747792
8
1


### 실습 7

In [15]:
# First 20 countries with employment data
countries = np.array([
    'Afghanistan', 'Albania', 'Algeria', 'Angola', 'Argentina',
    'Armenia', 'Australia', 'Austria', 'Azerbaijan', 'Bahamas',
    'Bahrain', 'Bangladesh', 'Barbados', 'Belarus', 'Belgium',
    'Belize', 'Benin', 'Bhutan', 'Bolivia',
    'Bosnia and Herzegovina'
])

# Employment data in 2007 for those 20 countries
employment = np.array([
    55.70000076,  51.40000153,  50.5       ,  75.69999695,
    58.40000153,  40.09999847,  61.5       ,  57.09999847,
    60.90000153,  66.59999847,  60.40000153,  68.09999847,
    66.90000153,  53.40000153,  48.59999847,  56.79999924,
    71.59999847,  58.40000153,  70.40000153,  41.20000076
])

#### 실습 7-1: 다음의 값을 출력하시오

* 3번째 나라의 이름
* 2~10번째 나라의 이름
* 17번째 이후의 나라의 이름

In [16]:
# 아래에 코드를 입력하시오.
print(countries[2])
print(countries[1:10])
print(countries[16:])

Algeria
['Albania' 'Algeria' 'Angola' 'Argentina' 'Armenia' 'Australia' 'Austria'
 'Azerbaijan' 'Bahamas']
['Benin' 'Bhutan' 'Bolivia' 'Bosnia and Herzegovina']


#### 실습 7-2: 다음과 같이 각 나라의 employment data를 출력하시오.
```
 Country Afghanistan has employment 55.70000076
 Country Albania has employment 51.40000153
 ...
```

In [17]:
# 아래의 코드를 완성하시오.
for i in range(len(countries)):
    country = countries[i]
    country_employment = employment[i]
    print('Country {} has employment {}'.format(country, country_employment))

Country Afghanistan has employment 55.70000076
Country Albania has employment 51.40000153
Country Algeria has employment 50.5
Country Angola has employment 75.69999695
Country Argentina has employment 58.40000153
Country Armenia has employment 40.09999847
Country Australia has employment 61.5
Country Austria has employment 57.09999847
Country Azerbaijan has employment 60.90000153
Country Bahamas has employment 66.59999847
Country Bahrain has employment 60.40000153
Country Bangladesh has employment 68.09999847
Country Barbados has employment 66.90000153
Country Belarus has employment 53.40000153
Country Belgium has employment 48.59999847
Country Belize has employment 56.79999924
Country Benin has employment 71.59999847
Country Bhutan has employment 58.40000153
Country Bolivia has employment 70.40000153
Country Bosnia and Herzegovina has employment 41.20000076


#### 실습 7-3: employment의 mean, std, max, mean 을 출력하시오.

In [18]:
# 아래에 코드를 입력하시오.
print(employment.mean())
print(employment.std())
print(employment.max())
print(employment.sum())

58.68500003850001
9.338269113687888
75.69999695
1173.70000077


In [19]:
# First 20 countries with school completion data
countries = np.array([
       'Algeria', 'Argentina', 'Armenia', 'Aruba', 'Austria','Azerbaijan',
       'Bahamas', 'Barbados', 'Belarus', 'Belgium', 'Belize', 'Bolivia',
       'Botswana', 'Brunei', 'Bulgaria', 'Burkina Faso', 'Burundi',
       'Cambodia', 'Cameroon', 'Cape Verde'
])

# Female school completion rate in 2007 for those 20 countries
female_completion = np.array([
    97.35583,   94.62379,   83.02998,   95.14321,   83.69019,
    98.49185,   99.88828,   95.43974,   92.11484,   91.54804,
    95.98029,   98.22902,   96.12179,   19.28105,   97.84627,
    29.07386,   38.41644,   90.70509,   51.7478 ,   95.45072
])

# Male school completion rate in 2007 for those 20 countries
male_completion = np.array([
     95.47622,   50.66476,   99.7926 ,   91.48936,   73.22096,
     97.80458,   83.81398,   88.11736,   93.55611,   87.76347,
     82.45714,   98.73953,   92.22388,   15.3892 ,   98.70502,
     37.00692,   45.39401,   91.22084,   62.42028,   90.66958
])

#### practice 7-4: 다음의 array는 선택된 20개 나라의 남성과 여성의 school completion data이다. 남녀의 completion 차가 가장 큰 나라를 찾는 함수를 작성하시오

In [21]:
def diff_completion(countries, female_completion, male_completion):
    diff = []
    for i in range(len(countries)):
        diff.append(abs(female_completion[i]-male_completion[i]))
    diff = np.array(diff)
    print(diff)
    return diff.max()

In [23]:
diff_completion(countries, female_completion, male_completion)

[ 1.87961 43.95903 16.76262  3.65385 10.46923  0.68727 16.0743   7.32238
  1.44127  3.78457 13.52315  0.51051  3.89791  3.89185  0.85875  7.93306
  6.97757  0.51575 10.67248  4.78114]


43.95903

## Pandas Basics

Pandas를 사용하기 위해서 다음과 같이 라이브러리를 import 한다. numpy도 주로 같이 사용하니 같이 import 한다.

In [24]:
import pandas as pd
import numpy as np

In [25]:
a = {'Name':['a','b','c','d','e'], 'Number':[1,2,3,4,5]}

In [26]:
df = pd.DataFrame(a)
df

Name  Number
0    a       1
1    b       2
2    c       3
3    d       4
4    e       5

In [27]:
df['Name']

0    a
1    b
2    c
3    d
4    e
Name: Name, dtype: object

In [28]:
df['Name'][2]

'c'

In [29]:
df['Color'] = ['red','blue','green','yellow','black']
df

Name  Number   Color
0    a       1     red
1    b       2    blue
2    c       3   green
3    d       4  yellow
4    e       5   black

In [30]:
del df['Name']
df

Number   Color
0       1     red
1       2    blue
2       3   green
3       4  yellow
4       5   black

Pandas는 read_csv()라는 CSV 파일을 읽어주는 함수를 제공한다.

(참고) Excel 파일을 읽는 함수도 제공한다. ExcelFile(), read_excel()
* df = pd.ExcelFile("dummydata.xlsx")
* df = pd.read_excel(open('your_xls_xlsx_filename','rb'), sheetname='Sheet 1')

In [31]:
data = pd.read_csv("data/weather_year.csv")

In [32]:
data

EDT  Max TemperatureF  Mean TemperatureF  Min TemperatureF  \
0    2012-3-10                56                 40                24   
1    2012-3-11                67                 49                30   
2    2012-3-12                71                 62                53   
3    2012-3-13                76                 63                50   
4    2012-3-14                80                 62                44   
5    2012-3-15                79                 69                58   
6    2012-3-16                75                 64                52   
7    2012-3-17                78                 62                46   
8    2012-3-18                80                 70                59   
9    2012-3-19                84                 72                59   
10   2012-3-20                84                 71                58   
11   2012-3-21                84                 71                57   
12   2012-3-22                81                 69                57   
13   2012-3-23                73                 64                55   
14   2012-3-24                65                 56                46   
15   2012-3-25                69                 61                52   
16   2012-3-26                63                 54                44   
17   2012-3-27                70                 54                38   
18   2012-3-28                77                 64                51   
19   2012-3-29                69                 58                46   
20   2012-3-30                81                 66                51   
21   2012-3-31                66                 58                49   
22    2012-4-1                79                 64                48   
23    2012-4-2                68                 62                55   
24    2012-4-3                84                 69                53   
25    2012-4-4                68                 62                55   
26    2012-4-5                61                 53                44   
27    2012-4-6                61                 48                35   
28    2012-4-7                67                 50                32   
29    2012-4-8                66                 56                45   
..         ...               ...                ...               ...   
336   2013-2-9                43                 33                23   
337  2013-2-10                52                 44                35   
338  2013-2-11                53                 45                36   
339  2013-2-12                45                 35                24   
340  2013-2-13                44                 37                30   
341  2013-2-14                52                 42                31   
342  2013-2-15                39                 30                21   
343  2013-2-16                29                 24                19   
344  2013-2-17                36                 24                12   
345  2013-2-18                58                 43                27   
346  2013-2-19                47                 35                23   
347  2013-2-20                28                 21                14   
348  2013-2-21                33                 24                15   
349  2013-2-22                37                 34                31   
350  2013-2-23                38                 31                23   
351  2013-2-24                45                 33                20   
352  2013-2-25                47                 36                24   
353  2013-2-26                44                 40                35   
354  2013-2-27                40                 36                32   
355  2013-2-28                37                 35                32   
356   2013-3-1                36                 34                31   
357   2013-3-2                32                 30                27   
358   2013-3-3                37                 28                19

In [33]:
len(data)

366

columns 명령어는 데이터의 column을 보여준다. 

In [34]:
data.columns

Index(['EDT', 'Max TemperatureF', 'Mean TemperatureF', 'Min TemperatureF',
       'Max Dew PointF', 'MeanDew PointF', 'Min DewpointF', 'Max Humidity',
       ' Mean Humidity', ' Min Humidity', ' Max Sea Level PressureIn',
       ' Mean Sea Level PressureIn', ' Min Sea Level PressureIn',
       ' Max VisibilityMiles', ' Mean VisibilityMiles', ' Min VisibilityMiles',
       ' Max Wind SpeedMPH', ' Mean Wind SpeedMPH', ' Max Gust SpeedMPH',
       'PrecipitationIn', ' CloudCover', ' Events', ' WindDirDegrees'],
      dtype='object')

In [35]:
len(data.columns)

23

특정 column을 읽어오려면 다음과 같이 쓴다.

In [36]:
data['EDT']

0      2012-3-10
1      2012-3-11
2      2012-3-12
3      2012-3-13
4      2012-3-14
5      2012-3-15
6      2012-3-16
7      2012-3-17
8      2012-3-18
9      2012-3-19
10     2012-3-20
11     2012-3-21
12     2012-3-22
13     2012-3-23
14     2012-3-24
15     2012-3-25
16     2012-3-26
17     2012-3-27
18     2012-3-28
19     2012-3-29
20     2012-3-30
21     2012-3-31
22      2012-4-1
23      2012-4-2
24      2012-4-3
25      2012-4-4
26      2012-4-5
27      2012-4-6
28      2012-4-7
29      2012-4-8
         ...    
336     2013-2-9
337    2013-2-10
338    2013-2-11
339    2013-2-12
340    2013-2-13
341    2013-2-14
342    2013-2-15
343    2013-2-16
344    2013-2-17
345    2013-2-18
346    2013-2-19
347    2013-2-20
348    2013-2-21
349    2013-2-22
350    2013-2-23
351    2013-2-24
352    2013-2-25
353    2013-2-26
354    2013-2-27
355    2013-2-28
356     2013-3-1
357     2013-3-2
358     2013-3-3
359     2013-3-4
360     2013-3-5
361     2013-3-6
362     2013-3-7
363     2013-3

어레이를 이용하여 여러개의 column을 불러올 수 있다.

In [37]:
data[['EDT', 'Max TemperatureF', 'Mean TemperatureF']]

EDT  Max TemperatureF  Mean TemperatureF
0    2012-3-10                56                 40
1    2012-3-11                67                 49
2    2012-3-12                71                 62
3    2012-3-13                76                 63
4    2012-3-14                80                 62
5    2012-3-15                79                 69
6    2012-3-16                75                 64
7    2012-3-17                78                 62
8    2012-3-18                80                 70
9    2012-3-19                84                 72
10   2012-3-20                84                 71
11   2012-3-21                84                 71
12   2012-3-22                81                 69
13   2012-3-23                73                 64
14   2012-3-24                65                 56
15   2012-3-25                69                 61
16   2012-3-26                63                 54
17   2012-3-27                70                 54
18   2012-3-28                77                 64
19   2012-3-29                69                 58
20   2012-3-30                81                 66
21   2012-3-31                66                 58
22    2012-4-1                79                 64
23    2012-4-2                68                 62
24    2012-4-3                84                 69
25    2012-4-4                68                 62
26    2012-4-5                61                 53
27    2012-4-6                61                 48
28    2012-4-7                67                 50
29    2012-4-8                66                 56
..         ...               ...                ...
336   2013-2-9                43                 33
337  2013-2-10                52                 44
338  2013-2-11                53                 45
339  2013-2-12                45                 35
340  2013-2-13                44                 37
341  2013-2-14                52                 42
342  2013-2-15                39                 30
343  2013-2-16                29                 24
344  2013-2-17                36                 24
345  2013-2-18                58                 43
346  2013-2-19                47                 35
347  2013-2-20                28                 21
348  2013-2-21                33                 24
349  2013-2-22                37                 34
350  2013-2-23                38                 31
351  2013-2-24                45                 33
352  2013-2-25                47                 36
353  2013-2-26                44                 40
354  2013-2-27                40                 36
355  2013-2-28                37                 35
356   2013-3-1                36                 34
357   2013-3-2                32                 30
358   2013-3-3                37                 28
359   2013-3-4                40                 30
360   2013-3-5                40                 35
361   2013-3-6                32                 31
362   2013-3-7                36                 32
363   2013-3-8                47                 35
364   2013-3-9                56                 45
365  2013-3-10                60                 52

[366 rows x 3 columns]

describe() 명령어는 기초 통계값을 보여준다. 숫자에만 적용된다.

In [38]:
data[['EDT', 'Max TemperatureF', 'Mean TemperatureF']].describe()

Max TemperatureF  Mean TemperatureF
count        366.000000         366.000000
mean          66.803279          55.683060
std           20.361247          18.436506
min           16.000000          11.000000
25%           51.000000          41.000000
50%           69.000000          59.000000
75%           84.000000          70.750000
max          106.000000          89.000000

만약 column 이름에 공백이 없다면, **```data.column_name```** 형태로 표현 가능하다.

In [39]:
data.EDT # only works when the column name has no space

0      2012-3-10
1      2012-3-11
2      2012-3-12
3      2012-3-13
4      2012-3-14
5      2012-3-15
6      2012-3-16
7      2012-3-17
8      2012-3-18
9      2012-3-19
10     2012-3-20
11     2012-3-21
12     2012-3-22
13     2012-3-23
14     2012-3-24
15     2012-3-25
16     2012-3-26
17     2012-3-27
18     2012-3-28
19     2012-3-29
20     2012-3-30
21     2012-3-31
22      2012-4-1
23      2012-4-2
24      2012-4-3
25      2012-4-4
26      2012-4-5
27      2012-4-6
28      2012-4-7
29      2012-4-8
         ...    
336     2013-2-9
337    2013-2-10
338    2013-2-11
339    2013-2-12
340    2013-2-13
341    2013-2-14
342    2013-2-15
343    2013-2-16
344    2013-2-17
345    2013-2-18
346    2013-2-19
347    2013-2-20
348    2013-2-21
349    2013-2-22
350    2013-2-23
351    2013-2-24
352    2013-2-25
353    2013-2-26
354    2013-2-27
355    2013-2-28
356     2013-3-1
357     2013-3-2
358     2013-3-3
359     2013-3-4
360     2013-3-5
361     2013-3-6
362     2013-3-7
363     2013-3

head()는 다섯개의 샘플 데이터만 보여준다.

In [40]:
data.head()

EDT  Max TemperatureF  Mean TemperatureF  Min TemperatureF  \
0  2012-3-10                56                 40                24   
1  2012-3-11                67                 49                30   
2  2012-3-12                71                 62                53   
3  2012-3-13                76                 63                50   
4  2012-3-14                80                 62                44   

   Max Dew PointF  MeanDew PointF  Min DewpointF  Max Humidity  \
0              24              20             16            74   
1              43              31             24            78   
2              59              55             43            90   
3              57              53             47            93   
4              58              52             43            93   

    Mean Humidity   Min Humidity       ...         Max VisibilityMiles  \
0              50             26       ...                          10   
1              53             28       ...                          10   
2              76             61       ...                          10   
3              66             38       ...                          10   
4              68             42       ...                          10   

    Mean VisibilityMiles   Min VisibilityMiles   Max Wind SpeedMPH  \
0                     10                    10                  13   
1                     10                    10                  22   
2                     10                     6                  24   
3                     10                     4                  16   
4                     10                    10                  16   

    Mean Wind SpeedMPH   Max Gust SpeedMPH  PrecipitationIn   CloudCover  \
0                    6                17.0             0.00            0   
1                    7                32.0                T            1   
2                   14                36.0             0.03            6   
3                    5                24.0             0.00            0   
4                    6                22.0             0.00            0   

    Events  WindDirDegrees  
0      NaN             138  
1     Rain             163  
2     Rain             190  
3      NaN             242  
4      NaN             202  

[5 rows x 23 columns]

data.head(num)는 원하는 수 num 만큼의 데이터를 보여준다.

In [41]:
data.head(10)

EDT  Max TemperatureF  Mean TemperatureF  Min TemperatureF  \
0  2012-3-10                56                 40                24   
1  2012-3-11                67                 49                30   
2  2012-3-12                71                 62                53   
3  2012-3-13                76                 63                50   
4  2012-3-14                80                 62                44   
5  2012-3-15                79                 69                58   
6  2012-3-16                75                 64                52   
7  2012-3-17                78                 62                46   
8  2012-3-18                80                 70                59   
9  2012-3-19                84                 72                59   

   Max Dew PointF  MeanDew PointF  Min DewpointF  Max Humidity  \
0              24              20             16            74   
1              43              31             24            78   
2              59              55             43            90   
3              57              53             47            93   
4              58              52             43            93   
5              61              58             53            90   
6              57              54             51           100   
7              60              54             46           100   
8              61              58             57            93   
9              58              56             50            90   

    Mean Humidity   Min Humidity       ...         Max VisibilityMiles  \
0              50             26       ...                          10   
1              53             28       ...                          10   
2              76             61       ...                          10   
3              66             38       ...                          10   
4              68             42       ...                          10   
5              69             48       ...                          10   
6              75             49       ...                          10   
7              78             56       ...                          10   
8              69             45       ...                          10   
9              66             42       ...                          10   

    Mean VisibilityMiles   Min VisibilityMiles   Max Wind SpeedMPH  \
0                     10                    10                  13   
1                     10                    10                  22   
2                     10                     6                  24   
3                     10                     4                  16   
4                     10                    10                  16   
5                     10                    10                  31   
6                     10                    10                  14   
7                      5                     0                  12   
8                     10                     9                  18   
9                     10                    10                  17   

    Mean Wind SpeedMPH   Max Gust SpeedMPH  PrecipitationIn   CloudCover  \
0                    6                17.0             0.00            0   
1                    7                32.0                T            1   
2                   14                36.0             0.03            6   
3                    5                24.0             0.00            0   
4                    6                22.0             0.00            0   
5                   10                41.0             0.04            3   
6                    5                20.0                T            2   
7                    5                17.0                T            3   
8                    8                25.0                T            2   
9                    6                23.0             0.00            1   

              Events  WindDirDegrees  
0                NaN             138  

tail() 함수는 head()와 반대로 맨 아래 다섯개의 데이터를 보여준다.

In [42]:
data[['EDT', 'Max TemperatureF', 'Mean TemperatureF']].tail()

EDT  Max TemperatureF  Mean TemperatureF
361   2013-3-6                32                 31
362   2013-3-7                36                 32
363   2013-3-8                47                 35
364   2013-3-9                56                 45
365  2013-3-10                60                 52

다시 columns 를 출력해보자.

In [43]:
data.columns

Index(['EDT', 'Max TemperatureF', 'Mean TemperatureF', 'Min TemperatureF',
       'Max Dew PointF', 'MeanDew PointF', 'Min DewpointF', 'Max Humidity',
       ' Mean Humidity', ' Min Humidity', ' Max Sea Level PressureIn',
       ' Mean Sea Level PressureIn', ' Min Sea Level PressureIn',
       ' Max VisibilityMiles', ' Mean VisibilityMiles', ' Min VisibilityMiles',
       ' Max Wind SpeedMPH', ' Mean Wind SpeedMPH', ' Max Gust SpeedMPH',
       'PrecipitationIn', ' CloudCover', ' Events', ' WindDirDegrees'],
      dtype='object')

In [44]:
data2 = data.copy()

columns는 단어 중간이나 앞부분에 공백이 있기도 하고 길어서 데이터 처리에 불편하다. columns를 다음과 같이 고쳐보자. column들의 이름이 바뀌었다.

In [45]:
data2.columns = ["date", "max_temp", "mean_temp", "min_temp", "max_dew",
                "mean_dew", "min_dew", "max_humidity", "mean_humidity",
                "min_humidity", "max_pressure", "mean_pressure",
                "min_pressure", "max_visibilty", "mean_visibility",
                "min_visibility", "max_wind", "mean_wind", "min_wind",
                "precipitation", "cloud_cover", "events", "wind_dir"]

In [46]:
data2.head()

date  max_temp  mean_temp  min_temp  max_dew  mean_dew  min_dew  \
0  2012-3-10        56         40        24       24        20       16   
1  2012-3-11        67         49        30       43        31       24   
2  2012-3-12        71         62        53       59        55       43   
3  2012-3-13        76         63        50       57        53       47   
4  2012-3-14        80         62        44       58        52       43   

   max_humidity  mean_humidity  min_humidity   ...     max_visibilty  \
0            74             50            26   ...                10   
1            78             53            28   ...                10   
2            90             76            61   ...                10   
3            93             66            38   ...                10   
4            93             68            42   ...                10   

   mean_visibility  min_visibility  max_wind  mean_wind  min_wind  \
0               10              10        13          6      17.0   
1               10              10        22          7      32.0   
2               10               6        24         14      36.0   
3               10               4        16          5      24.0   
4               10              10        16          6      22.0   

   precipitation  cloud_cover  events wind_dir  
0           0.00            0     NaN      138  
1              T            1    Rain      163  
2           0.03            6    Rain      190  
3           0.00            0     NaN      242  
4           0.00            0     NaN      202  

[5 rows x 23 columns]

In [47]:
data2.min_temp.head()

0    24
1    30
2    53
3    50
4    44
Name: min_temp, dtype: int64

In [48]:
data2.min_temp.std()

17.30114095238694

In [49]:
data2.std()

max_temp           20.361247
mean_temp          18.436506
min_temp           17.301141
max_dew            16.397178
mean_dew           16.829996
min_dew            17.479449
max_humidity        9.108438
mean_humidity       9.945591
min_humidity       15.360261
max_pressure        0.172189
mean_pressure       0.174112
min_pressure        0.182476
max_visibilty       0.073821
mean_visibility     1.875406
min_visibility      3.792219
max_wind            5.564329
mean_wind           3.200940
min_wind            8.131092
cloud_cover         2.707261
wind_dir           94.045080
dtype: float64

현재 우리 데이터의 각각의 row의 index는 0 ~ len(data)의 integer가 자동으로 지정되었다. 만약 데이터의 index를 다른 컬럼으로 (예를 들어 date) 지정하고 싶으면 index 명령어를 사용한다.

In [52]:
data2.index = data2.date
data2.head()

date  max_temp  mean_temp  min_temp  max_dew  mean_dew  \
date                                                                     
2012-3-10  2012-3-10        56         40        24       24        20   
2012-3-11  2012-3-11        67         49        30       43        31   
2012-3-12  2012-3-12        71         62        53       59        55   
2012-3-13  2012-3-13        76         63        50       57        53   
2012-3-14  2012-3-14        80         62        44       58        52   

           min_dew  max_humidity  mean_humidity  min_humidity   ...     \
date                                                            ...      
2012-3-10       16            74             50            26   ...      
2012-3-11       24            78             53            28   ...      
2012-3-12       43            90             76            61   ...      
2012-3-13       47            93             66            38   ...      
2012-3-14       43            93             68            42   ...      

           max_visibilty  mean_visibility  min_visibility  max_wind  \
date                                                                  
2012-3-10             10               10              10        13   
2012-3-11             10               10              10        22   
2012-3-12             10               10               6        24   
2012-3-13             10               10               4        16   
2012-3-14             10               10              10        16   

           mean_wind  min_wind  precipitation  cloud_cover  events wind_dir  
date                                                                         
2012-3-10          6      17.0           0.00            0     NaN      138  
2012-3-11          7      32.0              T            1    Rain      163  
2012-3-12         14      36.0           0.03            6    Rain      190  
2012-3-13          5      24.0           0.00            0     NaN      242  
2012-3-14          6      22.0           0.00            0     NaN      202  

[5 rows x 23 columns]

불필요한 컬럼을 삭제하기 위해서는 drop() 명령어를 사용한다. 위의 데이터에서는 date를 index로 지정하였기 때문에 데이터 테이블 내의 date는 필요없다. (현재 중복되어 있다) 

다음과 같이 삭제할 수 있다.

In [51]:
data2 = data2.drop(['date'])

KeyError: "['date'] not found in axis"

In [53]:
data2 = data2.drop(["date"], axis=1)
data2.columns

Index(['max_temp', 'mean_temp', 'min_temp', 'max_dew', 'mean_dew', 'min_dew',
       'max_humidity', 'mean_humidity', 'min_humidity', 'max_pressure',
       'mean_pressure', 'min_pressure', 'max_visibilty', 'mean_visibility',
       'min_visibility', 'max_wind', 'mean_wind', 'min_wind', 'precipitation',
       'cloud_cover', 'events', 'wind_dir'],
      dtype='object')

In [54]:
data2.head()


max_temp  mean_temp  min_temp  max_dew  mean_dew  min_dew  \
date                                                                   
2012-3-10        56         40        24       24        20       16   
2012-3-11        67         49        30       43        31       24   
2012-3-12        71         62        53       59        55       43   
2012-3-13        76         63        50       57        53       47   
2012-3-14        80         62        44       58        52       43   

           max_humidity  mean_humidity  min_humidity  max_pressure    ...     \
date                                                                  ...      
2012-3-10            74             50            26         30.53    ...      
2012-3-11            78             53            28         30.37    ...      
2012-3-12            90             76            61         30.13    ...      
2012-3-13            93             66            38         30.12    ...      
2012-3-14            93             68            42         30.15    ...      

           max_visibilty  mean_visibility  min_visibility  max_wind  \
date                                                                  
2012-3-10             10               10              10        13   
2012-3-11             10               10              10        22   
2012-3-12             10               10               6        24   
2012-3-13             10               10               4        16   
2012-3-14             10               10              10        16   

           mean_wind  min_wind  precipitation  cloud_cover events  wind_dir  
date                                                                         
2012-3-10          6      17.0           0.00            0    NaN       138  
2012-3-11          7      32.0              T            1   Rain       163  
2012-3-12         14      36.0           0.03            6   Rain       190  
2012-3-13          5      24.0           0.00            0    NaN       242  
2012-3-14          6      22.0           0.00            0    NaN       202  

[5 rows x 22 columns]

axis=1 은 column에 적용됨을 의미한다. 0은 row에, 1은 column에 적용된다고 생각하면 된다.

## Missing Values

이 데이터에는 missing value 가 있다. missing value 를 처리해보자.

In [55]:
empty = data2.isnull()
empty

max_temp  mean_temp  min_temp  max_dew  mean_dew  min_dew  \
date                                                                   
2012-3-10     False      False     False    False     False    False   
2012-3-11     False      False     False    False     False    False   
2012-3-12     False      False     False    False     False    False   
2012-3-13     False      False     False    False     False    False   
2012-3-14     False      False     False    False     False    False   
2012-3-15     False      False     False    False     False    False   
2012-3-16     False      False     False    False     False    False   
2012-3-17     False      False     False    False     False    False   
2012-3-18     False      False     False    False     False    False   
2012-3-19     False      False     False    False     False    False   
2012-3-20     False      False     False    False     False    False   
2012-3-21     False      False     False    False     False    False   
2012-3-22     False      False     False    False     False    False   
2012-3-23     False      False     False    False     False    False   
2012-3-24     False      False     False    False     False    False   
2012-3-25     False      False     False    False     False    False   
2012-3-26     False      False     False    False     False    False   
2012-3-27     False      False     False    False     False    False   
2012-3-28     False      False     False    False     False    False   
2012-3-29     False      False     False    False     False    False   
2012-3-30     False      False     False    False     False    False   
2012-3-31     False      False     False    False     False    False   
2012-4-1      False      False     False    False     False    False   
2012-4-2      False      False     False    False     False    False   
2012-4-3      False      False     False    False     False    False   
2012-4-4      False      False     False    False     False    False   
2012-4-5      False      False     False    False     False    False   
2012-4-6      False      False     False    False     False    False   
2012-4-7      False      False     False    False     False    False   
2012-4-8      False      False     False    False     False    False   
...             ...        ...       ...      ...       ...      ...   
2013-2-9      False      False     False    False     False    False   
2013-2-10     False      False     False    False     False    False   
2013-2-11     False      False     False    False     False    False   
2013-2-12     False      False     False    False     False    False   
2013-2-13     False      False     False    False     False    False   
2013-2-14     False      False     False    False     False    False   
2013-2-15     False      False     False    False     False    False   
2013-2-16     False      False     False    False     False    False   
2013-2-17     False      False     False    False     False    False   
2013-2-18     False      False     False    False     False    False   
2013-2-19     False      False     False    False     False    False   
2013-2-20     False      False     False    False     False    False   
2013-2-21     False      False     False    False     False    False   
2013-2-22     False      False     False    False     False    False   
2013-2-23     False      False     False    False     False    False   
2013-2-24     False      False     False    False     False    False   
2013-2-25     False      False     False    False     False    False   
2013-2-26     False      False     False    False     False    False   
2013-2-27     False      False     False    False     False    False   
2013-2-28     False      False     False    False     False    False   
2013-3-1      False      False     False    False     False    False   
2013-3-2      False      False     False    False     False    False   
2013-3-3      False      False     False    False  

In [56]:
empty.events.head(10)

date
2012-3-10     True
2012-3-11    False
2012-3-12    False
2012-3-13     True
2012-3-14     True
2012-3-15    False
2012-3-16     True
2012-3-17    False
2012-3-18    False
2012-3-19     True
Name: events, dtype: bool

In [57]:
data2.events.head(10)

date
2012-3-10                  NaN
2012-3-11                 Rain
2012-3-12                 Rain
2012-3-13                  NaN
2012-3-14                  NaN
2012-3-15    Rain-Thunderstorm
2012-3-16                  NaN
2012-3-17     Fog-Thunderstorm
2012-3-18                 Rain
2012-3-19                  NaN
Name: events, dtype: object

In [58]:
len(empty[empty.events == True])

204

missing data를 처리하는 한가지 방법은 해당 데이터 포인트를 삭제하는 것. pandas에는 dropna()라는 함수가 있다. 이 함수를 사용하면 데이터가 입력되지 않은 row, 즉 위의 데이터에서 events가 NaN인 경우 모두 삭제를 한다. 그러나, 이렇게 삭제를 하게되면 전체 366개의 데이터 중 162개 밖에 남지 않는다. 따라서 이렇게 missing data가 많은 경우 삭제하는 것 보다는 다르게 처리하는 편이 낫다. 우리는 NaN 대신 빈 문자열을 채워 넣으려고 한다. 이때 사용하는 함수는 fillna()이다.

In [59]:
data2.events = data2.events.fillna("")
data2.events.head(10)

date
2012-3-10                     
2012-3-11                 Rain
2012-3-12                 Rain
2012-3-13                     
2012-3-14                     
2012-3-15    Rain-Thunderstorm
2012-3-16                     
2012-3-17     Fog-Thunderstorm
2012-3-18                 Rain
2012-3-19                     
Name: events, dtype: object

## Accessing Rows

In [60]:
data2.iloc[0]

max_temp              56
mean_temp             40
min_temp              24
max_dew               24
mean_dew              20
min_dew               16
max_humidity          74
mean_humidity         50
min_humidity          26
max_pressure       30.53
mean_pressure      30.45
min_pressure       30.34
max_visibilty         10
mean_visibility       10
min_visibility        10
max_wind              13
mean_wind              6
min_wind              17
precipitation       0.00
cloud_cover            0
events                  
wind_dir             138
Name: 2012-3-10, dtype: object

In [61]:
data2.iloc[3:9]

max_temp  mean_temp  min_temp  max_dew  mean_dew  min_dew  \
date                                                                   
2012-3-13        76         63        50       57        53       47   
2012-3-14        80         62        44       58        52       43   
2012-3-15        79         69        58       61        58       53   
2012-3-16        75         64        52       57        54       51   
2012-3-17        78         62        46       60        54       46   
2012-3-18        80         70        59       61        58       57   

           max_humidity  mean_humidity  min_humidity  max_pressure    ...     \
date                                                                  ...      
2012-3-13            93             66            38         30.12    ...      
2012-3-14            93             68            42         30.15    ...      
2012-3-15            90             69            48         30.13    ...      
2012-3-16           100             75            49         30.14    ...      
2012-3-17           100             78            56         30.15    ...      
2012-3-18            93             69            45         30.14    ...      

           max_visibilty  mean_visibility  min_visibility  max_wind  \
date                                                                  
2012-3-13             10               10               4        16   
2012-3-14             10               10              10        16   
2012-3-15             10               10              10        31   
2012-3-16             10               10              10        14   
2012-3-17             10                5               0        12   
2012-3-18             10               10               9        18   

           mean_wind  min_wind  precipitation  cloud_cover             events  \
date                                                                            
2012-3-13          5      24.0           0.00            0                      
2012-3-14          6      22.0           0.00            0                      
2012-3-15         10      41.0           0.04            3  Rain-Thunderstorm   
2012-3-16          5      20.0              T            2                      
2012-3-17          5      17.0              T            3   Fog-Thunderstorm   
2012-3-18          8      25.0              T            2               Rain   

           wind_dir  
date                 
2012-3-13       242  
2012-3-14       202  
2012-3-15       209  
2012-3-16       169  
2012-3-17       162  
2012-3-18       197  

[6 rows x 22 columns]

In [62]:
data2.loc['2012-3-16']

max_temp              75
mean_temp             64
min_temp              52
max_dew               57
mean_dew              54
min_dew               51
max_humidity         100
mean_humidity         75
min_humidity          49
max_pressure       30.14
mean_pressure       30.1
min_pressure       30.07
max_visibilty         10
mean_visibility       10
min_visibility        10
max_wind              14
mean_wind              5
min_wind              20
precipitation          T
cloud_cover            2
events                  
wind_dir             169
Name: 2012-3-16, dtype: object

In [63]:
num_rain = 0
for i in range(len(data2)):
    if data2.events[i] == 'Rain':
        num_rain += 1

"Days with rain: {0}".format(num_rain)  

'Days with rain: 69'

## Filtering

Filtering은 데이터를 살펴보고 분석을 할 때 가장 좋은 방법이다. Filtering을 하기 위한 방법들을 살펴보자.

In [64]:
freezing_days = data2[data2.max_temp <= 32]
freezing_days

max_temp  mean_temp  min_temp  max_dew  mean_dew  min_dew  \
date                                                                    
2012-11-24        31         26        21       20        18       15   
2012-12-21        29         26        22       25        19       15   
2012-12-29        32         28        23       28        25       16   
2012-12-30        31         18         4       21        12        1   
2013-1-1          32         26        20       31        25       16   
2013-1-2          25         13         1       13         7       -2   
2013-1-4          31         24        17       23        20       14   
2013-1-14         26         22        18       19        16       12   
2013-1-21         24         18        12       16         7       -4   
2013-1-22         18         12         5        0        -3       -5   
2013-1-23         31         20         9       17         8        0   
2013-1-24         26         19        11       14         4       -5   
2013-1-25         30         25        20       18        12        0   
2013-1-31         31         21        10       22        14       -2   
2013-2-1          16         11         5        6         0       -3   
2013-2-2          30         22        14       22        18        7   
2013-2-3          27         19        11       21        15        8   
2013-2-16         29         24        19       16        12        7   
2013-2-20         28         21        14       13         9        6   
2013-3-2          32         30        27       27        22       19   
2013-3-6          32         31        29       27        26       25   

            max_humidity  mean_humidity  min_humidity  max_pressure    ...     \
date                                                                   ...      
2012-11-24            81             72            63         30.30    ...      
2012-12-21            85             74            63         30.21    ...      
2012-12-29            92             80            68         30.29    ...      
2012-12-30            92             75            58         30.47    ...      
2013-1-1              92             83            74         30.20    ...      
2013-1-2              92             71            50         30.29    ...      
2013-1-4              92             75            58         30.39    ...      
2013-1-14             88             81            74         30.46    ...      
2013-1-21             74             59            43         30.32    ...      
2013-1-22             69             56            43         30.44    ...      
2013-1-23             73             56            39         30.39    ...      
2013-1-24             73             50            27         30.60    ...      
2013-1-25             74             57            39         30.35    ...      
2013-1-31             81             67            53         30.19    ...      
2013-2-1              76             65            53         30.47    ...      
2013-2-2              84             71            58         30.30    ...      
2013-2-3              92             80            67         30.15    ...      
2013-2-16             84             63            42         30.17    ...      
2013-2-20             73             58            42         30.35    ...      
2013-3-2              85             72            59         30.14    ...      
2013-3-6              92             85            78         30.31    ...      

            max_visibilty  mean_visibility  min_visibility  max_wind  \
date                                                                   
2012-11-24             10               10               9         9   
2012-12-21             10                5               0        25   
2012-12-29             10                3               0        18   
2012-12-30             10                6               0        15   
2013-1-1                9           

In [65]:
freezing_days[freezing_days.min_temp >= 20]

max_temp  mean_temp  min_temp  max_dew  mean_dew  min_dew  \
date                                                                    
2012-11-24        31         26        21       20        18       15   
2012-12-21        29         26        22       25        19       15   
2012-12-29        32         28        23       28        25       16   
2013-1-1          32         26        20       31        25       16   
2013-1-25         30         25        20       18        12        0   
2013-3-2          32         30        27       27        22       19   
2013-3-6          32         31        29       27        26       25   

            max_humidity  mean_humidity  min_humidity  max_pressure    ...     \
date                                                                   ...      
2012-11-24            81             72            63         30.30    ...      
2012-12-21            85             74            63         30.21    ...      
2012-12-29            92             80            68         30.29    ...      
2013-1-1              92             83            74         30.20    ...      
2013-1-25             74             57            39         30.35    ...      
2013-3-2              85             72            59         30.14    ...      
2013-3-6              92             85            78         30.31    ...      

            max_visibilty  mean_visibility  min_visibility  max_wind  \
date                                                                   
2012-11-24             10               10               9         9   
2012-12-21             10                5               0        25   
2012-12-29             10                3               0        18   
2013-1-1                9                5               2        14   
2013-1-25              10                8               1        16   
2013-3-2               10                9               2        20   
2013-3-6               10                5               1        22   

            mean_wind  min_wind  precipitation  cloud_cover    events  \
date                                                                    
2012-11-24          4      14.0           0.00            4             
2012-12-21         14      39.0           0.02            7  Fog-Snow   
2012-12-29          9      29.0           0.20            8  Fog-Snow   
2013-1-1            5      15.0              T            8             
2013-1-25           7      21.0           0.02            6      Snow   
2013-3-2            8      23.0              T            8      Snow   
2013-3-6            9      33.0           0.04            8      Snow   

            wind_dir  
date                  
2012-11-24       270  
2012-12-21       285  
2012-12-29       308  
2013-1-1         353  
2013-1-25        192  
2013-3-2         351  
2013-3-6         314  

[7 rows x 22 columns]

In [66]:
data2[(data2.max_temp <= 32) & (data2.min_temp >= 20)]

max_temp  mean_temp  min_temp  max_dew  mean_dew  min_dew  \
date                                                                    
2012-11-24        31         26        21       20        18       15   
2012-12-21        29         26        22       25        19       15   
2012-12-29        32         28        23       28        25       16   
2013-1-1          32         26        20       31        25       16   
2013-1-25         30         25        20       18        12        0   
2013-3-2          32         30        27       27        22       19   
2013-3-6          32         31        29       27        26       25   

            max_humidity  mean_humidity  min_humidity  max_pressure    ...     \
date                                                                   ...      
2012-11-24            81             72            63         30.30    ...      
2012-12-21            85             74            63         30.21    ...      
2012-12-29            92             80            68         30.29    ...      
2013-1-1              92             83            74         30.20    ...      
2013-1-25             74             57            39         30.35    ...      
2013-3-2              85             72            59         30.14    ...      
2013-3-6              92             85            78         30.31    ...      

            max_visibilty  mean_visibility  min_visibility  max_wind  \
date                                                                   
2012-11-24             10               10               9         9   
2012-12-21             10                5               0        25   
2012-12-29             10                3               0        18   
2013-1-1                9                5               2        14   
2013-1-25              10                8               1        16   
2013-3-2               10                9               2        20   
2013-3-6               10                5               1        22   

            mean_wind  min_wind  precipitation  cloud_cover    events  \
date                                                                    
2012-11-24          4      14.0           0.00            4             
2012-12-21         14      39.0           0.02            7  Fog-Snow   
2012-12-29          9      29.0           0.20            8  Fog-Snow   
2013-1-1            5      15.0              T            8             
2013-1-25           7      21.0           0.02            6      Snow   
2013-3-2            8      23.0              T            8      Snow   
2013-3-6            9      33.0           0.04            8      Snow   

            wind_dir  
date                  
2012-11-24       270  
2012-12-21       285  
2012-12-29       308  
2013-1-1         353  
2013-1-25        192  
2013-3-2         351  
2013-3-6         314  

[7 rows x 22 columns]

필터를 변수 형태로 만들어 놓고 사용하는 것도 물론 가능하다.

In [67]:
# max_temp 가 32보다 작거나 같은 경우의 필터를 생성
temp_max = data2.max_temp <= 32
temp_max

date
2012-3-10    False
2012-3-11    False
2012-3-12    False
2012-3-13    False
2012-3-14    False
2012-3-15    False
2012-3-16    False
2012-3-17    False
2012-3-18    False
2012-3-19    False
2012-3-20    False
2012-3-21    False
2012-3-22    False
2012-3-23    False
2012-3-24    False
2012-3-25    False
2012-3-26    False
2012-3-27    False
2012-3-28    False
2012-3-29    False
2012-3-30    False
2012-3-31    False
2012-4-1     False
2012-4-2     False
2012-4-3     False
2012-4-4     False
2012-4-5     False
2012-4-6     False
2012-4-7     False
2012-4-8     False
             ...  
2013-2-9     False
2013-2-10    False
2013-2-11    False
2013-2-12    False
2013-2-13    False
2013-2-14    False
2013-2-15    False
2013-2-16     True
2013-2-17    False
2013-2-18    False
2013-2-19    False
2013-2-20     True
2013-2-21    False
2013-2-22    False
2013-2-23    False
2013-2-24    False
2013-2-25    False
2013-2-26    False
2013-2-27    False
2013-2-28    False
2013-3-1     False
2013-3-

In [221]:
data2[temp_max]

max_temp  mean_temp  min_temp  max_dew  mean_dew  min_dew  \
date                                                                    
2012-11-24        31         26        21       20        18       15   
2012-12-21        29         26        22       25        19       15   
2012-12-29        32         28        23       28        25       16   
2012-12-30        31         18         4       21        12        1   
2013-1-1          32         26        20       31        25       16   
2013-1-2          25         13         1       13         7       -2   
2013-1-4          31         24        17       23        20       14   
2013-1-14         26         22        18       19        16       12   
2013-1-21         24         18        12       16         7       -4   
2013-1-22         18         12         5        0        -3       -5   
2013-1-23         31         20         9       17         8        0   
2013-1-24         26         19        11       14         4       -5   
2013-1-25         30         25        20       18        12        0   
2013-1-31         31         21        10       22        14       -2   
2013-2-1          16         11         5        6         0       -3   
2013-2-2          30         22        14       22        18        7   
2013-2-3          27         19        11       21        15        8   
2013-2-16         29         24        19       16        12        7   
2013-2-20         28         21        14       13         9        6   
2013-3-2          32         30        27       27        22       19   
2013-3-6          32         31        29       27        26       25   

            max_humidity  mean_humidity  min_humidity  max_pressure    ...     \
date                                                                   ...      
2012-11-24            81             72            63         30.30    ...      
2012-12-21            85             74            63         30.21    ...      
2012-12-29            92             80            68         30.29    ...      
2012-12-30            92             75            58         30.47    ...      
2013-1-1              92             83            74         30.20    ...      
2013-1-2              92             71            50         30.29    ...      
2013-1-4              92             75            58         30.39    ...      
2013-1-14             88             81            74         30.46    ...      
2013-1-21             74             59            43         30.32    ...      
2013-1-22             69             56            43         30.44    ...      
2013-1-23             73             56            39         30.39    ...      
2013-1-24             73             50            27         30.60    ...      
2013-1-25             74             57            39         30.35    ...      
2013-1-31             81             67            53         30.19    ...      
2013-2-1              76             65            53         30.47    ...      
2013-2-2              84             71            58         30.30    ...      
2013-2-3              92             80            67         30.15    ...      
2013-2-16             84             63            42         30.17    ...      
2013-2-20             73             58            42         30.35    ...      
2013-3-2              85             72            59         30.14    ...      
2013-3-6              92             85            78         30.31    ...      

            max_visibilty  mean_visibility  min_visibility  max_wind  \
date                                                                   
2012-11-24             10               10               9         9   
2012-12-21             10                5               0        25   
2012-12-29             10                3               0        18   
2012-12-30             10                6               0        15   
2013-1-1                9           

In [69]:
# max_temp 가 20보다 크거나 같은 경우의 필터를 생성
temp_min = data2.min_temp >= 20
temp_min

date
2012-3-10     True
2012-3-11     True
2012-3-12     True
2012-3-13     True
2012-3-14     True
2012-3-15     True
2012-3-16     True
2012-3-17     True
2012-3-18     True
2012-3-19     True
2012-3-20     True
2012-3-21     True
2012-3-22     True
2012-3-23     True
2012-3-24     True
2012-3-25     True
2012-3-26     True
2012-3-27     True
2012-3-28     True
2012-3-29     True
2012-3-30     True
2012-3-31     True
2012-4-1      True
2012-4-2      True
2012-4-3      True
2012-4-4      True
2012-4-5      True
2012-4-6      True
2012-4-7      True
2012-4-8      True
             ...  
2013-2-9      True
2013-2-10     True
2013-2-11     True
2013-2-12     True
2013-2-13     True
2013-2-14     True
2013-2-15     True
2013-2-16    False
2013-2-17    False
2013-2-18     True
2013-2-19     True
2013-2-20    False
2013-2-21    False
2013-2-22     True
2013-2-23     True
2013-2-24     True
2013-2-25     True
2013-2-26     True
2013-2-27     True
2013-2-28     True
2013-3-1      True
2013-3-

In [70]:
temp_min & temp_max

date
2012-3-10    False
2012-3-11    False
2012-3-12    False
2012-3-13    False
2012-3-14    False
2012-3-15    False
2012-3-16    False
2012-3-17    False
2012-3-18    False
2012-3-19    False
2012-3-20    False
2012-3-21    False
2012-3-22    False
2012-3-23    False
2012-3-24    False
2012-3-25    False
2012-3-26    False
2012-3-27    False
2012-3-28    False
2012-3-29    False
2012-3-30    False
2012-3-31    False
2012-4-1     False
2012-4-2     False
2012-4-3     False
2012-4-4     False
2012-4-5     False
2012-4-6     False
2012-4-7     False
2012-4-8     False
             ...  
2013-2-9     False
2013-2-10    False
2013-2-11    False
2013-2-12    False
2013-2-13    False
2013-2-14    False
2013-2-15    False
2013-2-16    False
2013-2-17    False
2013-2-18    False
2013-2-19    False
2013-2-20    False
2013-2-21    False
2013-2-22    False
2013-2-23    False
2013-2-24    False
2013-2-25    False
2013-2-26    False
2013-2-27    False
2013-2-28    False
2013-3-1     False
2013-3-

In [71]:
temp_min | temp_max

date
2012-3-10     True
2012-3-11     True
2012-3-12     True
2012-3-13     True
2012-3-14     True
2012-3-15     True
2012-3-16     True
2012-3-17     True
2012-3-18     True
2012-3-19     True
2012-3-20     True
2012-3-21     True
2012-3-22     True
2012-3-23     True
2012-3-24     True
2012-3-25     True
2012-3-26     True
2012-3-27     True
2012-3-28     True
2012-3-29     True
2012-3-30     True
2012-3-31     True
2012-4-1      True
2012-4-2      True
2012-4-3      True
2012-4-4      True
2012-4-5      True
2012-4-6      True
2012-4-7      True
2012-4-8      True
             ...  
2013-2-9      True
2013-2-10     True
2013-2-11     True
2013-2-12     True
2013-2-13     True
2013-2-14     True
2013-2-15     True
2013-2-16     True
2013-2-17    False
2013-2-18     True
2013-2-19     True
2013-2-20     True
2013-2-21    False
2013-2-22     True
2013-2-23     True
2013-2-24     True
2013-2-25     True
2013-2-26     True
2013-2-27     True
2013-2-28     True
2013-3-1      True
2013-3-

두개의 필터를 이용해서 새로운 필터를 만들어 둘 수 있다.

In [72]:
temp_both = temp_min & temp_max

In [73]:
data2[data2.events == 'Rain']

max_temp  mean_temp  min_temp  max_dew  mean_dew  min_dew  \
date                                                                    
2012-3-11         67         49        30       43        31       24   
2012-3-12         71         62        53       59        55       43   
2012-3-18         80         70        59       61        58       57   
2012-3-22         81         69        57       63        57       51   
2012-3-29         69         58        46       45        39       35   
2012-4-5          61         53        44       43        35       30   
2012-4-8          66         56        45       46        34       20   
2012-4-16         73         61        48       62        52       43   
2012-4-19         76         59        42       48        42       40   
2012-4-20         73         58        43       53        47       41   
2012-4-24         64         50        35       47        34       26   
2012-4-27         63         50        36       30        25       21   
2012-4-29         68         57        46       56        48       40   
2012-5-8          74         65        55       64        56       44   
2012-5-9          68         58        48       49        46       41   
2012-6-1          62         56        50       59        53       47   
2012-6-2          70         60        49       56        48       44   
2012-6-4          68         62        56       63        59       54   
2012-7-14         83         77        70       73        70       67   
2012-7-26         94         86        77       72        69       64   
2012-8-3          86         77        68       72        69       67   
2012-8-5          92         77        62       74        67       55   
2012-8-13         81         70        59       64        60       55   
2012-8-17         80         68        55       65        60       51   
2012-8-21         80         66        52       60        55       49   
2012-8-27         87         77        67       72        69       62   
2012-8-31         86         79        71       72        70       63   
2012-9-1          84         79        74       73        72       71   
2012-9-2          76         74        72       73        72       70   
2012-9-8          72         64        56       67        58       48   
...              ...        ...       ...      ...       ...      ...   
2012-10-18        62         55        48       55        45       28   
2012-10-19        54         50        46       45        43       34   
2012-10-20        53         47        40       46        44       39   
2012-10-23        77         70        63       56        51       48   
2012-10-26        57         48        39       52        41       35   
2012-10-30        44         42        39       34        30       28   
2012-11-3         41         38        34       38        35       29   
2012-11-6         51         38        25       39        31       23   
2012-11-7         47         40        32       39        38       31   
2012-11-11        67         60        53       53        44       41   
2012-11-23        53         41        28       51        34       17   
2012-12-2         63         60        56       59        54       47   
2012-12-3         71         66        61       59        56       49   
2012-12-4         65         51        37       58        49       37   
2012-12-8         52         49        46       52        46       39   
2012-12-15        56         48        39       52        41       27   
2012-12-18        48         42        36       38        34       33   
2012-12-19        46         39        32       37        33       28   
2013-1-9          51         41        30       40        35       29   
2013-1-10         47         38        29       45        36       27   
2013-1-12         66         61        55       61        55       47   
2013-1-13         65         45        25       61        37       20 

## Grouping

groupby() 함수는 dataframe에서 같은 값을 갖는 데이터 포인트를 찾아서 묶어 준다.

예를 들어 cloud_cover 데이터를 추출하면 다음과 같이 0-8까지 값이 기록되어 있다.

In [74]:
data2.cloud_cover

date
2012-3-10    0
2012-3-11    1
2012-3-12    6
2012-3-13    0
2012-3-14    0
2012-3-15    3
2012-3-16    2
2012-3-17    3
2012-3-18    2
2012-3-19    1
2012-3-20    0
2012-3-21    1
2012-3-22    3
2012-3-23    7
2012-3-24    5
2012-3-25    6
2012-3-26    2
2012-3-27    0
2012-3-28    2
2012-3-29    2
2012-3-30    4
2012-3-31    4
2012-4-1     4
2012-4-2     3
2012-4-3     0
2012-4-4     6
2012-4-5     3
2012-4-6     0
2012-4-7     1
2012-4-8     2
            ..
2013-2-9     2
2013-2-10    7
2013-2-11    4
2013-2-12    1
2013-2-13    4
2013-2-14    4
2013-2-15    5
2013-2-16    4
2013-2-17    1
2013-2-18    4
2013-2-19    8
2013-2-20    2
2013-2-21    5
2013-2-22    8
2013-2-23    1
2013-2-24    0
2013-2-25    1
2013-2-26    8
2013-2-27    8
2013-2-28    8
2013-3-1     8
2013-3-2     8
2013-3-3     3
2013-3-4     4
2013-3-5     8
2013-3-6     8
2013-3-7     8
2013-3-8     1
2013-3-9     2
2013-3-10    1
Name: cloud_cover, Length: 366, dtype: int64

In [75]:
data2.cloud_cover.unique()

array([0, 1, 6, 3, 2, 7, 5, 4, 8])

In [76]:
for d in data2.groupby("cloud_cover"):
    print(d)

# 결과는 tuple 형태로 출력된다. 
# d[0] -> cloud_cover level
# d[1] -> 각 cloud_cover level이 포함된 data row
# 여기서 loop는 각 data row의 iteration을 가져오는게 아니라 group의 iteration을 가져온다.

(0,             max_temp  mean_temp  min_temp  max_dew  mean_dew  min_dew  \
date                                                                    
2012-3-10         56         40        24       24        20       16   
2012-3-13         76         63        50       57        53       47   
2012-3-14         80         62        44       58        52       43   
2012-3-20         84         71        58       56        54       50   
2012-3-27         70         54        38       42        32       25   
2012-4-3          84         69        53       60        51       41   
2012-4-6          61         48        35       35        31       28   
2012-4-9          67         55        43       36        26       18   
2012-4-10         54         44        34       29        23       15   
2012-4-12         61         45        28       34        28       24   
2012-4-17         66         53        40       47        43       39   
2012-4-18         71         54        37      

In [86]:
df2 = data2.groupby("cloud_cover")
df2.groups

{0: Index(['2012-3-10', '2012-3-13', '2012-3-14', '2012-3-20', '2012-3-27',
        '2012-4-3', '2012-4-6', '2012-4-9', '2012-4-10', '2012-4-12',
        ...
        '2012-12-13', '2012-12-14', '2012-12-23', '2013-1-2', '2013-1-7',
        '2013-1-8', '2013-1-18', '2013-1-19', '2013-1-26', '2013-2-24'],
       dtype='object', name='date', length=104),
 1: Index(['2012-3-11', '2012-3-19', '2012-3-21', '2012-4-7', '2012-4-11',
        '2012-4-15', '2012-4-23', '2012-5-2', '2012-5-10', '2012-5-12',
        '2012-5-20', '2012-5-30', '2012-6-3', '2012-6-6', '2012-6-8',
        '2012-6-10', '2012-6-12', '2012-6-18', '2012-6-22', '2012-7-2',
        '2012-7-4', '2012-7-6', '2012-7-12', '2012-7-17', '2012-7-18',
        '2012-7-26', '2012-7-28', '2012-7-30', '2012-7-31', '2012-8-8',
        '2012-8-12', '2012-8-19', '2012-8-21', '2012-8-25', '2012-8-26',
        '2012-8-31', '2012-9-23', '2012-10-8', '2012-10-24', '2012-10-25',
        '2012-11-9', '2012-11-21', '2012-11-22', '2012-11-30', '20

출력되는 tuple 값을 이용해서 각 그룹의 평균온도(mean_temp)의 평균을 다음과 같이 구할 수 있다.

In [87]:
cover_temps = {}
for cover, cover_data in data2.groupby("cloud_cover"):
    cover_temps[cover] = cover_data.mean_temp.mean()
cover_temps

{0: 59.73076923076923,
 1: 61.41509433962264,
 2: 59.72727272727273,
 3: 58.0625,
 4: 51.5,
 5: 50.827586206896555,
 6: 57.72727272727273,
 7: 46.5,
 8: 40.90909090909091}

한개 이상의 컬럼을 이용해서 데이터를 groupby 할 수 있다. 이 경우, 두개의 컬럼의 값이 tuple 로 먼저 출력되고, 그 다음에 data row 가 출력된다.

```예) a = ((1,2),3)
a[0] => (1,2)
a[1] => 3
a[0][0] = 1
```

In [88]:
for d in data2.groupby(["cloud_cover", "events"]):
    print(d)

((0, ''),             max_temp  mean_temp  min_temp  max_dew  mean_dew  min_dew  \
date                                                                    
2012-3-10         56         40        24       24        20       16   
2012-3-13         76         63        50       57        53       47   
2012-3-14         80         62        44       58        52       43   
2012-3-20         84         71        58       56        54       50   
2012-3-27         70         54        38       42        32       25   
2012-4-3          84         69        53       60        51       41   
2012-4-6          61         48        35       35        31       28   
2012-4-9          67         55        43       36        26       18   
2012-4-10         54         44        34       29        23       15   
2012-4-12         61         45        28       34        28       24   
2012-4-17         66         53        40       47        43       39   
2012-4-18         71         54        37

[4 rows x 22 columns])
((3, 'Snow'),           max_temp  mean_temp  min_temp  max_dew  mean_dew  min_dew  \
date                                                                  
2013-3-3        37         28        19       22        18       13   

          max_humidity  mean_humidity  min_humidity  max_pressure    ...     \
date                                                                 ...      
2013-3-3            84             63            41         30.15    ...      

          max_visibilty  mean_visibility  min_visibility  max_wind  mean_wind  \
date                                                                            
2013-3-3             10                9               2        10          4   

          min_wind  precipitation  cloud_cover events  wind_dir  
date                                                             
2013-3-3      14.0              T            3   Snow       291  

[1 rows x 22 columns])
((4, ''),             max_temp  mean_temp  mi

[6 rows x 22 columns])


In [92]:
df2 = data2.groupby(["cloud_cover", "events"])
df2.groups

{(0,
  ''): Index(['2012-3-10', '2012-3-13', '2012-3-14', '2012-3-20', '2012-3-27',
        '2012-4-3', '2012-4-6', '2012-4-9', '2012-4-10', '2012-4-12',
        '2012-4-17', '2012-4-18', '2012-4-26', '2012-5-3', '2012-5-6',
        '2012-5-11', '2012-5-15', '2012-5-16', '2012-5-17', '2012-5-18',
        '2012-5-19', '2012-5-23', '2012-5-24', '2012-5-26', '2012-5-27',
        '2012-5-28', '2012-6-7', '2012-6-9', '2012-6-13', '2012-6-14',
        '2012-6-15', '2012-6-16', '2012-6-19', '2012-6-20', '2012-6-23',
        '2012-6-24', '2012-6-26', '2012-6-27', '2012-6-28', '2012-6-29',
        '2012-6-30', '2012-7-5', '2012-7-10', '2012-7-11', '2012-7-16',
        '2012-7-21', '2012-7-22', '2012-7-23', '2012-7-24', '2012-7-25',
        '2012-7-29', '2012-8-1', '2012-8-2', '2012-8-6', '2012-8-7',
        '2012-8-11', '2012-8-18', '2012-8-23', '2012-8-24', '2012-8-28',
        '2012-8-29', '2012-8-30', '2012-9-10', '2012-9-11', '2012-9-15',
        '2012-9-19', '2012-9-24', '2012-9-30', '2012

In [93]:
for (cover, events), group_data in data2.groupby(["cloud_cover", "events"]):
    print("Cover: {}, Events: {}, Count: {}".format(cover, events, len(group_data)))

Cover: 0, Events: , Count: 99
Cover: 0, Events: Fog, Count: 2
Cover: 0, Events: Rain, Count: 2
Cover: 0, Events: Thunderstorm, Count: 1
Cover: 1, Events: , Count: 35
Cover: 1, Events: Fog, Count: 5
Cover: 1, Events: Fog-Rain, Count: 1
Cover: 1, Events: Rain, Count: 4
Cover: 1, Events: Rain-Thunderstorm, Count: 2
Cover: 1, Events: Thunderstorm, Count: 6
Cover: 2, Events: , Count: 20
Cover: 2, Events: Fog, Count: 1
Cover: 2, Events: Rain, Count: 5
Cover: 2, Events: Rain-Thunderstorm, Count: 4
Cover: 2, Events: Snow, Count: 1
Cover: 2, Events: Thunderstorm, Count: 2
Cover: 3, Events: , Count: 12
Cover: 3, Events: Fog, Count: 2
Cover: 3, Events: Fog-Rain-Thunderstorm, Count: 3
Cover: 3, Events: Fog-Thunderstorm, Count: 1
Cover: 3, Events: Rain, Count: 9
Cover: 3, Events: Rain-Thunderstorm, Count: 4
Cover: 3, Events: Snow, Count: 1
Cover: 4, Events: , Count: 16
Cover: 4, Events: Fog, Count: 3
Cover: 4, Events: Fog-Rain, Count: 2
Cover: 4, Events: Fog-Rain-Thunderstorm, Count: 2
Cover: 4, Ev

In [94]:
for (cover, events), group_data in data2.groupby(["cloud_cover", "events"]):
    
    print(group_data['max_temp'])

date
2012-3-10     56
2012-3-13     76
2012-3-14     80
2012-3-20     84
2012-3-27     70
2012-4-3      84
2012-4-6      61
2012-4-9      67
2012-4-10     54
2012-4-12     61
2012-4-17     66
2012-4-18     71
2012-4-26     75
2012-5-3      86
2012-5-6      86
2012-5-11     75
2012-5-15     81
2012-5-16     79
2012-5-17     75
2012-5-18     82
2012-5-19     87
2012-5-23     80
2012-5-24     86
2012-5-26     91
2012-5-27     92
2012-5-28     92
2012-6-7      83
2012-6-9      87
2012-6-13     82
2012-6-14     86
              ..
2012-10-11    60
2012-10-12    62
2012-10-16    69
2012-10-21    71
2012-10-27    54
2012-10-29    55
2012-11-2     55
2012-11-4     49
2012-11-5     50
2012-11-10    73
2012-11-13    43
2012-11-14    44
2012-11-15    47
2012-11-16    53
2012-11-17    57
2012-11-18    61
2012-11-25    48
2012-11-28    45
2012-11-29    51
2012-12-5     51
2012-12-12    45
2012-12-13    50
2012-12-14    54
2012-12-23    48
2013-1-2      25
2013-1-8      45
2013-1-18     42
2013-1-19

## DataFrame 병합하기

In [95]:
df1 = pd.DataFrame({'A': ['A0', 'A1', 'A2', 'A3'], 
                    'B': ['B0', 'B1', 'B2', 'B3'],
                    'C': ['C0', 'C1', 'C2', 'C3'],
                    'D': ['D0', 'D1', 'D2', 'D3']},
                   index=[0, 1, 2, 3])

df2 = pd.DataFrame({'A': ['A4', 'A5', 'A6', 'A7'],
                    'B': ['B4', 'B5', 'B6', 'B7'],
                    'C': ['C4', 'C5', 'C6', 'C7'],
                    'D': ['D4', 'D5', 'D6', 'D7']},
                   index=[4, 5, 6, 7])

df3 = pd.DataFrame({'A': ['A8', 'A9', 'A10', 'A11'],
                    'B': ['B8', 'B9', 'B10', 'B11'],
                    'C': ['C8', 'C9', 'C10', 'C11'],
                    'D': ['D8', 'D9', 'D10', 'D11']},
                   index=[8, 9, 10, 11])

In [96]:
df1

A   B   C   D
0  A0  B0  C0  D0
1  A1  B1  C1  D1
2  A2  B2  C2  D2
3  A3  B3  C3  D3

In [97]:
df2

A   B   C   D
4  A4  B4  C4  D4
5  A5  B5  C5  D5
6  A6  B6  C6  D6
7  A7  B7  C7  D7

In [98]:
df3

A    B    C    D
8    A8   B8   C8   D8
9    A9   B9   C9   D9
10  A10  B10  C10  D10
11  A11  B11  C11  D11

In [99]:
result = pd.concat([df1, df2, df3])
result

A    B    C    D
0    A0   B0   C0   D0
1    A1   B1   C1   D1
2    A2   B2   C2   D2
3    A3   B3   C3   D3
4    A4   B4   C4   D4
5    A5   B5   C5   D5
6    A6   B6   C6   D6
7    A7   B7   C7   D7
8    A8   B8   C8   D8
9    A9   B9   C9   D9
10  A10  B10  C10  D10
11  A11  B11  C11  D11

In [100]:
df4 = pd.DataFrame({'B': ['B2', 'B3', 'B6', 'B7'], 
                    'D': ['D2', 'D3', 'D6', 'D7'],
                    'F': ['F2', 'F3', 'F6', 'F7']},
                   index=[2, 3, 6, 7])

result = pd.concat([df1, df4], axis=1)

In [101]:
df1

A   B   C   D
0  A0  B0  C0  D0
1  A1  B1  C1  D1
2  A2  B2  C2  D2
3  A3  B3  C3  D3

In [102]:
df4

B   D   F
2  B2  D2  F2
3  B3  D3  F3
6  B6  D6  F6
7  B7  D7  F7

In [103]:
result

A    B    C    D    B    D    F
0   A0   B0   C0   D0  NaN  NaN  NaN
1   A1   B1   C1   D1  NaN  NaN  NaN
2   A2   B2   C2   D2   B2   D2   F2
3   A3   B3   C3   D3   B3   D3   F3
6  NaN  NaN  NaN  NaN   B6   D6   F6
7  NaN  NaN  NaN  NaN   B7   D7   F7

In [104]:
result = pd.concat([df1, df4], axis=1, join='inner')
result

A   B   C   D   B   D   F
2  A2  B2  C2  D2  B2  D2  F2
3  A3  B3  C3  D3  B3  D3  F3

In [105]:
result = pd.concat([df1, df4], axis=1, join_axes=[df1.index])
result

A   B   C   D    B    D    F
0  A0  B0  C0  D0  NaN  NaN  NaN
1  A1  B1  C1  D1  NaN  NaN  NaN
2  A2  B2  C2  D2   B2   D2   F2
3  A3  B3  C3  D3   B3   D3   F3

In [106]:
result = pd.concat([df1, df4], ignore_index=True)
result

/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  """Entry point for launching an IPython kernel.


A   B    C   D    F
0   A0  B0   C0  D0  NaN
1   A1  B1   C1  D1  NaN
2   A2  B2   C2  D2  NaN
3   A3  B3   C3  D3  NaN
4  NaN  B2  NaN  D2   F2
5  NaN  B3  NaN  D3   F3
6  NaN  B6  NaN  D6   F6
7  NaN  B7  NaN  D7   F7

In [107]:
left = pd.DataFrame({'A': ['A0', 'A1', 'A2', 'A3'],
                     'B': ['B0', 'B1', 'B2', 'B3'],
                     'key': ['K0', 'K4', 'K2', 'K3']})

right = pd.DataFrame({'C': ['C0', 'C1', 'C2', 'C3'],
                      'D': ['D0', 'D1', 'D2', 'D3'],
                      'key': ['K0', 'K1', 'K2', 'K3']})

In [108]:
left

A   B key
0  A0  B0  K0
1  A1  B1  K4
2  A2  B2  K2
3  A3  B3  K3

In [109]:
right

C   D key
0  C0  D0  K0
1  C1  D1  K1
2  C2  D2  K2
3  C3  D3  K3

In [110]:
pd.merge(left, right, on='key')

A   B key   C   D
0  A0  B0  K0  C0  D0
1  A2  B2  K2  C2  D2
2  A3  B3  K3  C3  D3

In [111]:
pd.merge(left, right, how='left', on='key')

A   B key    C    D
0  A0  B0  K0   C0   D0
1  A1  B1  K4  NaN  NaN
2  A2  B2  K2   C2   D2
3  A3  B3  K3   C3   D3

In [112]:
pd.merge(left, right, how='right', on='key')

A    B key   C   D
0   A0   B0  K0  C0  D0
1   A2   B2  K2  C2  D2
2   A3   B3  K3  C3  D3
3  NaN  NaN  K1  C1  D1

In [113]:
pd.merge(right, left, how='left', on='key')

C   D key    A    B
0  C0  D0  K0   A0   B0
1  C1  D1  K1  NaN  NaN
2  C2  D2  K2   A2   B2
3  C3  D3  K3   A3   B3

In [114]:
pd.merge(left, right, how='outer', on='key')
# how = 'outer' Nan값까지 반환 

A    B key    C    D
0   A0   B0  K0   C0   D0
1   A1   B1  K4  NaN  NaN
2   A2   B2  K2   C2   D2
3   A3   B3  K3   C3   D3
4  NaN  NaN  K1   C1   D1

In [115]:
pd.merge(left, right, how='inner', on='key')

A   B key   C   D
0  A0  B0  K0  C0  D0
1  A2  B2  K2  C2  D2
2  A3  B3  K3  C3  D3

## Export Data

to_csv() 함수는 데이터프레임을 CSV형태로 저장한다.

In [116]:
data2.to_csv("data/weather-mod.csv")

## 실습

#### practice 7-5
* 주어진 employment.csv파일은 header가 없는 파일이다. 불러온 후 column의 header를 추가한다.
* 국가명을 인덱스로 지정한다.
* 불필요한 컬럼이 있으면 삭제한다.
* 상위 5개를 출력한다.

#### practice 7-6
* 다음 2개의 파일도 불러와서 employment.csv와 같은 방식으로 편집한다.
* life_expectancy.csv
* gdp_per_capita.csv

#### practice 7-7
* 세개의 데이터프레임을 하나로 합치자.
* 데이터프레임을 합칠 때는 merge, join 등의 개념이 사용된다.
* http://pandas.pydata.org/pandas-docs/stable/merging.html 를 참고하자.
* (힌트) concat 이 사용된다.

#### practice 7-8
* gdp 상위 10개 국가의 평균과 하위 10개 국가의 리스트와 값을 구하고 그리고 평균의 차이를 구해보자.